### PR TITLE
feat: add retry mechanism for OAuth 429

### PR DESCRIPTION
**Pull request by Kirdock**
#### **Why** this PR?
There may be 429 for every request to platform, which could break the whole Terraform plan/apply

#### **What** has changed?
429 returned by OAuth token requests are now retried

#### **How** does it do it?
This adds special handling for OAuth requests in case of 429. It now retries based on the set retry headers with a fallback to an exponential backoff with a maximum retry amount of 10

#### How is it **tested**?
Tests added and integration tests are still working

#### How does it affect **users**?
Less OAuth requests and less likely to run into 429

## Notes
V2 of https://github.com/dynatrace-oss/terraform-provider-dynatrace/pull/1041
Still, this one (with retry) doesn't solve the cause (too many not needed requests) and could lead to increasing execution times by a factor more than 10 compared to the cache.

**Issue:** CA-18762

### CODE REVIEW

File: dynatrace/rest/retry_transport.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+		}
+		if t, err := http.ParseTime(ra); err == nil {
+			if d := time.Until(t); d > 0 {
+				return d
+			}
+		}
+	}
+	// exponential backoff: defaultBackoff * 2^attempt
+	return time.Duration(float64(defaultBackoff) * math.Pow(2, float64(attempt)))
+}
```
Comment (arthurpitman) [inline dynatrace/rest/retry_transport.go]: I would split this function into two:
- `getRetryAfterHeaderAsSleepTime(resp *http.Response) time.Duration`
- `computeBackoffSleepTime(attempt int) time.Duration`

File: dynatrace/rest/retry_transport.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+		// Restore the body for each attempt.
+		if bodyBytes != nil {
+			req.Body = io.NopCloser(bytes.NewReader(bodyBytes))
+		}
+
+		resp, err = t.Transport.RoundTrip(req)
+		if err != nil {
+			return nil, err
+		}
+		if resp.StatusCode != http.StatusTooManyRequests {
```
Comment (arthurpitman) [inline dynatrace/rest/retry_transport.go]: I was wondering if this is the only thing we should retry. What about `503`?

File: dynatrace/rest/retry_transport.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+			return resp, nil
+		}
+
+		if attempt == maxRetries {
+			break
+		}
+
+		wait := backoff(resp, attempt)
+		resp.Body.Close()
```
Comment (arthurpitman) [inline dynatrace/rest/retry_transport.go]: Do we also need to ensure the body is read to completion?

File: dynatrace/api/iam/iam_client.go [hunk 4]
```diff
@@ -58,7 +58,7 @@ func NewIAMClient(a Authenticator) IAMClient {
 		ClientSecret: a.ClientSecret(),
 		TokenURL:     a.TokenURL(),
 	}
-	httpClient := auth.NewOAuthClient(context.Background(), &oauthConfig)
+	httpClient := auth.NewOAuthClient(rest.NewContextWithOAuthRetryClient(context.Background()), &oauthConfig)
```
Comment (tomazpu) [inline dynatrace/api/iam/iam_client.go:61]: I can see that even before we used an empty context. The question I ask myself is, since we're starting to implement a retry that can go on for a longer time(2560s), we should use a non-empty ctx, so if Terraform cancels the process, the cleanup happens earlier. Also, maybe we want to set the max time with context?

Comment (tomazpu) [reply dynatrace/api/iam/iam_client.go:61]: I understand, you did not introduce this. But I think we should fix it and also inject ctx like in the platform client.

File: dynatrace/rest/retry_transport.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+const (
+	maxRetries     = 10
+	defaultBackoff = 5 * time.Second
+)
+
+type RetryTransport struct {
+	Transport http.RoundTripper
+}
+
+func (t *RetryTransport) RoundTrip(req *http.Request) (*http.Response, error) {
```
Comment (tomazpu) [inline dynatrace/rest/retry_transport.go:107]: Should we log every time we retry?

File: dynatrace/rest/retry_transport.go (hunk 2 above)
Comment (Kirdock) [reply dynatrace/rest/retry_transport.go]: We haven't had any issues with SSO so far and 429 would be the only one that we would expect.
So I would say we don't need the 503 and can implement it if needed

File: dynatrace/rest/retry_transport.go (hunk 3 above)
Comment (Kirdock) [reply dynatrace/rest/retry_transport.go]: We don't need to read the body if we simply do a retry. It's enough to just close it.

File: dynatrace/rest/retry_transport.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+		if err != nil {
+			return nil, err
+		}
+		if resp.StatusCode != http.StatusTooManyRequests {
+			return resp, nil
+		}
+
+		if attempt == maxRetries {
+			break
+		}
```
Comment (tomazpu) [inline dynatrace/rest/retry_transport.go]: ` if resp.StatusCode != http.StatusTooManyRequests || attempt == maxRetries {                                                   
             return resp, nil                                                                                                          
}  `

Comment (tomazpu) [reply dynatrace/rest/retry_transport.go]: I would combine both exit conditions

File: dynatrace/rest/retry_transport.go [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+	"math"
+	"net/http"
+	"strconv"
+	"time"
+
+	"golang.org/x/oauth2"
+)
+
+const (
+	maxRetries     = 10
```
Comment (tomazpu) [inline dynatrace/rest/retry_transport.go]: If they're settable we could have a faster test

File: dynatrace/api/iam/iam_client.go (hunk 4 above)
Comment (Kirdock) [reply dynatrace/api/iam/iam_client.go:61]: This would require quite a bit of refactoring and affect several resources, and another point is that potential caching would not work because Terraform cancels the context for each resource.

File: dynatrace/rest/retry_transport.go (hunk 6 above)
Comment (Kirdock) [reply dynatrace/rest/retry_transport.go]: Could do that but it would slightly reduce the code coverage as the last return can never be reached

File: dynatrace/rest/retry_transport.go (hunk 2 above)
Comment (Laubi) [reply dynatrace/rest/retry_transport.go]: Maybe we should just handle it right now, so we don't have to worry about it in the future 🤔 
503 (or generally speaking, 5xx errors) can always happen. They shouldn't, but they do

File: dynatrace/rest/retry_transport.go (hunk 3 above)
Comment (Laubi) [reply dynatrace/rest/retry_transport.go]: This is not entirely true
https://pkg.go.dev/net/http#Client.Do
> If the returned error is nil, the [Response](https://pkg.go.dev/net/http#Response) will contain a non-nil Body which the user is expected to close. If the Body is not both read to EOF and closed, the [Client](https://pkg.go.dev/net/http#Client)'s underlying [RoundTripper](https://pkg.go.dev/net/http#RoundTripper) (typically [Transport](https://pkg.go.dev/net/http#Transport)) may not be able to re-use a persistent TCP connection to the server for a subsequent "keep-alive" request.

so this technically works, but is quite bad. the underlaying tcp connection cannot be reused.
1. Kills the existing connection
2. Forces a new TCP handshake
3. Forces a new TLS handshake

Please use a pattern like 
```go
io.Copy(io.Discard, resp.Body)
resp.Body.Close()
```

File: dynatrace/rest/retry_transport.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+			return time.Duration(seconds) * time.Second
+		}
+		if t, err := http.ParseTime(ra); err == nil {
+			if d := time.Until(t); d > 0 {
+				return d
+			}
+		}
+	}
+	// exponential backoff: defaultBackoff * 2^attempt
+	return time.Duration(float64(defaultBackoff) * math.Pow(2, float64(attempt)))
```
Comment (Laubi) [inline dynatrace/rest/retry_transport.go]: At 10 retries this is a backoff of 85 minutes, note sure if we should wait that long. Maybe we should go with 1.5 and only 5 retries?

File: dynatrace/rest/retry_transport.go [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+		time.Sleep(wait)
+	}
+
+	return resp, nil
+}
+
+// backoff determines how long to wait before the next retry.
+// It first checks the Retry-After header; if absent it falls back to exponential backoff.
+func backoff(resp *http.Response, attempt int) time.Duration {
+	if ra := resp.Header.Get("Retry-After"); ra != "" {
```
Comment (Laubi) [inline dynatrace/rest/retry_transport.go]: We should set an upper bound, right? Also, we should add tests for edge cases, like 0 and negative values. Just to ensure that if the API returns invalid values, we handle them gracefully.

File: dynatrace/rest/retry_transport.go [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+		}
+	}
+
+	var resp *http.Response
+	var err error
+
+	for attempt := 0; attempt <= maxRetries; attempt++ {
+		// Restore the body for each attempt.
+		if bodyBytes != nil {
+			req.Body = io.NopCloser(bytes.NewReader(bodyBytes))
```
Comment (Laubi) [inline dynatrace/rest/retry_transport.go]: https://pkg.go.dev/net/http#RoundTripper

> // RoundTrip should not modify the request, except for
>	// consuming and closing the Request's Body. RoundTrip may
>	// read fields of the request in a separate goroutine. Callers
>	// should not mutate or reuse the request until the Response's
>	// Body has been closed.

Clone the request per attempt:
```go
clone := req.Clone(req.Context())
clone.Body = io.NopCloser(bytes.NewReader(bodyBytes))
resp, err = t.Transport.RoundTrip(clone)
```

File: dynatrace/rest/retry_transport.go (hunk 8 above)
Comment (Laubi) [inline dynatrace/rest/retry_transport.go]: nit: Maybe we should add some jitter?

File: dynatrace/rest/retry_transport.go [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+		}
+	}
+	// exponential backoff: defaultBackoff * 2^attempt
+	return time.Duration(float64(defaultBackoff) * math.Pow(2, float64(attempt)))
+}
+
+func NewContextWithOAuthRetryClient(ctx context.Context) context.Context {
+	return context.WithValue(ctx, oauth2.HTTPClient, &http.Client{
+		Transport: &RetryTransport{
+			Transport: http.DefaultTransport,
```
Comment (Laubi) [inline dynatrace/rest/retry_transport.go]: Why have the transport as a field if you always set it to defaultTransport?

File: dynatrace/rest/retry_transport.go [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+		req.Body.Close()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var resp *http.Response
+	var err error
+
+	for attempt := 0; attempt <= maxRetries; attempt++ {
```
Comment (Laubi) [inline dynatrace/rest/retry_transport.go]: We need to adhere to the context cancellation. 

the context may be canceled at any time during the retry, so we need to react to it. 
Possible solution is selecting whatever happens first:
```go
select {
case <-time.After(wait):
case <-req.Context().Done():
    resp.Body.Close()
    return nil, req.Context().Err()
}
```

**Review by Laubi — CHANGES_REQUESTED** [review]

File: dynatrace/rest/retry_transport.go (hunk 7 above)
Comment (Laubi) [reply dynatrace/rest/retry_transport.go]: I think we can use synctest: https://go.dev/blog/synctest

File: dynatrace/rest/retry_transport.go [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+	"golang.org/x/oauth2"
+)
+
+const (
+	maxRetries     = 10
+	defaultBackoff = 5 * time.Second
+)
+
+type RetryTransport struct {
+	Transport http.RoundTripper
```
Comment (Laubi) [inline dynatrace/rest/retry_transport.go]: Should this be public?
If yes, it should 100% be nil-checked

File: dynatrace/rest/retry_transport.go (hunk 6 above)
Comment (Laubi) [reply dynatrace/rest/retry_transport.go]: Why can it never be reached?

Comment (Kirdock) [reply dynatrace/rest/retry_transport.go]: It either returns on not 429 or max-attempts reached, which is inside the for-loop.

File: dynatrace/rest/retry_transport.go (hunk 11 above)
Comment (Kirdock) [reply dynatrace/rest/retry_transport.go]: Adjusted to not set it to default but still have the field to easily override it for tests

File: dynatrace/rest/retry_transport.go (hunk 13 above)
Comment (Kirdock) [reply dynatrace/rest/retry_transport.go]: Testing related. Nil check added

File: dynatrace/api/iam/iam_client.go (hunk 4 above)
Comment (Kirdock) [reply dynatrace/api/iam/iam_client.go:61]: Decided to do this in a follow-up 👍

File: dynatrace/rest/retry_transport.go [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+	defaultBaseBackoff = 5 * time.Second
+	defaultMaxBackoff  = 60 * time.Second
+)
+
+// RetryTransport is an http.RoundTripper that automatically retries requests
+// when the server returns HTTP 429 (Too Many Requests) or HTTP 503 (Service Unavailable).
+// MaxRetries, BaseBackoff, and MaxBackoff can be set to non-zero values to override the
+// defaults, which is useful in tests to keep execution time short.
+type RetryTransport struct {
+	// Transport is the underlying RoundTripper. If nil, http.DefaultTransport is used.
```
Comment (tomazpu) [inline dynatrace/rest/retry_transport.go]: NIT: I am ok with the comment above the struct, but the comments for the fields give me no value, except that the code is harder to read, in my personal opinion.

Comment (tomazpu) [reply dynatrace/rest/retry_transport.go]: If others agree i would ask you to only leave relevant comments that add value.

**Review by tomazpu — DISMISSED** [review]

**Review by tomazpu — DISMISSED** [review]

File: dynatrace/api/iam/iam_client.go (hunk 4 above)
Comment (Kirdock) [reply dynatrace/api/iam/iam_client.go:61]: Follow-up: https://github.com/dynatrace-oss/terraform-provider-dynatrace/pull/1048

File: dynatrace/rest/retry_transport.go [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+// computeBackoffSleepTime returns the exponential back-off duration for the given attempt
+// using a 1.5× multiplier (base × 1.5^attempt), with up to 10%
+// random jitter to avoid thundering-herd problems.
+func computeBackoffSleepTime(base time.Duration, attempt int) time.Duration {
+	backoff := time.Duration(float64(base) * math.Pow(1.5, float64(attempt)))
+	if backoff > 0 {
+		// Add up to 10% random jitter.
+		backoff += time.Duration(rand.Int64N(int64(backoff)/10 + 1))
+	}
+	return backoff
```
Comment (arthurpitman) [inline dynatrace/rest/retry_transport.go]:
```suggestion
	// random number between 1.0 and 1.1 to add up to 10% jitter
	scale := 1 + 0.1*rand.Float64()
	return time.Duration(float64(base) * scale * math.Pow(1.5, float64(attempt)))
```

**Review by arthurpitman — DISMISSED** [review]

**Review by Laubi — APPROVED** [review]
